theories/datatypes/MSet.ec

```diff
@@ -0,0 +1,689 @@
+require import Core Int List StdRing StdOrder FMap FSet StdBigop.
+(*---*) import IntOrder.
+
+
+(* -------------------------------------------------------------------- *)
+(* Finite multisets are represented as lists up to permutation          *)
+
+type 'a mset.
+
+op elems  : 'a mset -> 'a list.
+op oflist : 'a list -> 'a mset.
+
+axiom elemsK: cancel elems<:'a> oflist.
+axiom oflistK (s : 'a list): perm_eq s (elems (oflist s)).
+
+axiom mset_eq (s1 s2 : 'a mset):
+  (perm_eq (elems s1) (elems s2)) => (s1 = s2).
+
+(* -------------------------------------------------------------------- *)
+
+lemma perm_eq_oflistP (s1 s2 : 'a list) :
+  oflist s1 = oflist s2 <=> perm_eq s1 s2.
+proof.
+split. 
+- move=> oflist_eq; rewrite (perm_eq_trans _ _ _ (oflistK s1)) oflist_eq.
+  by rewrite perm_eq_sym oflistK.
+move=> peq_s1s2; apply/mset_eq.
+apply (perm_eq_trans s1 _ _); 1: rewrite perm_eq_sym oflistK.
+rewrite (perm_eq_trans s2 _ _ peq_s1s2) oflistK.
+qed.
+
+(* -------------------------------------------------------------------- *)
+op [opaque] mult ['a]  (x: 'a) (s: 'a mset): int = count (pred1 x) (elems s).
+
+abbrev "_.[_]" (s: 'a mset) (x: 'a) = mult x s.
+
+lemma oflist_mult (s : 'a list):
+  forall x, (oflist s).[x] = count (pred1 x) s.
+proof. move => x; rewrite/mult; apply/perm_eqP/perm_eq_sym/oflistK. qed.
+
+lemma mult_ge0 s (x: 'a): 0 <= s.[x] by rewrite/mult; apply count_ge0.
+
+hint exact : mult_ge0.
+
+lemma mset_eqP (s1 s2 : 'a mset):
+  (s1 = s2) <=> (forall x, s1.[x] = s2.[x]).
+proof.
+split=> // h; apply/mset_eq.
+rewrite/mult in h.
+by rewrite perm_eqP_pred1.
+qed.
+
+(* -------------------------------------------------------------------- *)
+op [opaque] mem ['a] (s : 'a mset) (x : 'a) = mem (elems s) x.
+lemma memE (s : 's mset) x: mem s x = mem (elems s) x by rewrite /mem.
+
+abbrev (\in)    (z : 'a) (s : 'a mset) =  mem s z.
+abbrev (\notin) (z : 'a) (s : 'a mset) = !mem s z.
+
+lemma mult_in_ge1 s (x: 'a): x \in s <=> 1 <= mult x s. 
+proof. rewrite /mult; have := mem_count x (elems s); rewrite -memE /#. qed.
+
+lemma mult0 s (x: 'a): x \notin s <=> mult x s = 0 by smt(mult_ge0 mult_in_ge1).
+
+lemma mem_oflist (s : 'a list):
+  forall x, x \in (oflist s) <=> x \in s.
+proof.
+move=> x; rewrite !memE (perm_eq_mem _ s) //.
+by rewrite perm_eq_sym oflistK.
+qed.
+
+(* -------------------------------------------------------------------- *)
+op selems (s: 'a mset): 'a fset = oflist (elems s).
+
+lemma mem_selems s (e: 'a): e \in selems s <=> e \in s.
+proof. by rewrite /selems mem_oflist memE. qed.
+
+op offset' (s: 'a fset): 'a mset = oflist (elems s).
+
+lemma offset_mult (s: 'a fset) x: (offset' s).[x] = b2i (x \in s).
+proof. 
+rewrite /mult /multisetify.
+have/perm_eqP_pred1<-:= (oflistK (elems s)).
+by rewrite count_uniq_mem ?uniq_elems /mem.
+qed.
+
+(* -------------------------------------------------------------------- *)
+
+op tofmap (s: 'a mset): ('a, int) fmap = offsetmap (fun e => s.[e]) (selems s).
```

Comment:
Consider making a lot more things `opaque`.

Reply:
I've added `opaque` to everything that touches the internals rather than things defined in the theory. That makes `rem`, `insert`, `\subset` and `\proper` exceptions.